lib/semantics/expression.h

```diff
@@ -311,24 +311,31 @@ class ExpressionAnalyzer {
   MaybeExpr TopLevelChecks(DataRef &&);
   std::optional<Expr<SubscriptInteger>> GetSubstringBound(
       const std::optional<parser::ScalarIntExpr> &);
-  std::optional<ProcedureDesignator> AnalyzeProcedureComponentRef(
-      const parser::ProcComponentRef &);
-  std::optional<ActualArgument> AnalyzeActualArgument(const parser::Expr &);
 
   struct CalleeAndArguments {
     ProcedureDesignator procedureDesignator;
     ActualArguments arguments;
   };
 
+  std::optional<CalleeAndArguments> AnalyzeProcedureComponentRef(
+      const parser::ProcComponentRef &, ActualArguments &&);
+  std::optional<ActualArgument> AnalyzeActualArgument(const parser::Expr &);
+
   MaybeExpr AnalyzeCall(const parser::Call &, bool isSubroutine);
   std::optional<ActualArguments> AnalyzeArguments(
       const parser::Call &, bool isSubroutine);
-  std::optional<CalleeAndArguments> Procedure(
-      const parser::ProcedureDesignator &, ActualArguments &);
+  std::optional<characteristics::Procedure> CheckCall(
+      parser::CharBlock, const ProcedureDesignator &, ActualArguments &);
+
+  std::optional<CalleeAndArguments> GetCalleeAndArguments(
+      const parser::ProcedureDesignator &, ActualArguments &&,
+      bool isSubroutine);
+
+  void CheckForBadRecursion(const semantics::Symbol &, parser::CharBlock);
   bool EnforceTypeConstraint(parser::CharBlock, const MaybeExpr &, TypeCategory,
       bool defaultKind = false);
-  MaybeExpr MakeFunctionRef(ProcedureDesignator &&, ActualArguments &&);
-  MaybeExpr MakeFunctionRef(CalleeAndArguments &&);
+  MaybeExpr MakeFunctionRef(ProcedureDesignator &&, ActualArguments &&,
+      parser::CharBlock callSite = {});
```

Comment:
It bugs me that `MakeFunctionRef` and `CheckForBadRecursion` have the `CharBlock` specifying the location at the end of the arguments whereas `CheckCall` and `EnforceTypeConstraint` have it at the beginning.

Reply:
Which one do you prefer?

Reply:
Most functions that have a `CharBlock` argument that represents a source location (e.g. the constructor for `Message`) has it first, so that would be my preference.

Reply:
Ok.  I'll implement the default value with an additional overload.